File: implementations/react/src/patterns/behavioral/state/closed-state.ts
import type { TicketContext } from "./context";
import { TicketState } from "./state";

export class ClosedState extends TicketState {
  private ctx: TicketContext;
  constructor(ctx: TicketContext) {
    super();
    this.ctx = ctx;
  }
  handleRequest() {
    if (!this.ctx.allDone()) {
      const remain = this.ctx.tasks.filter(t => !t.done).map(t => t.title).join(", ");
      throw new Error(`Cannot close. Remaining tasks: ${remain || "unknown"}`);
    }
    if (!this.ctx.closedAt) {
      this.ctx.closedAt = new Date();
    } else {
    }
  }
}

File: implementations/flutter/design_patterns_flutter/lib/patterns/behavioral/state/closed_state.dart
import 'package:design_patterns_flutter/patterns/behavioral/state/context.dart';
import 'package:design_patterns_flutter/patterns/behavioral/state/state.dart';

class ClosedState extends TicketState {
  final TicketContext ctx;
  ClosedState(this.ctx);

  @override
  void handleRequest() {
    if (!ctx.allDone()) {
      final remain = ctx.tasks
          .where((t) => !t.done)
          .map((t) => t.title)
          .join(', ');
      throw Exception(
        'Cannot close. Remaining tasks: ${remain.isEmpty ? 'unknown' : remain}',
      );
    }
    ctx.closedAt ??= DateTime.now();
  }
}
